### DISCUSSION TITLE
How to get num of bytes fuse_reply_data really sent?

### DISCUSSION THREAD

**Discussion by rockeet** [libfuse/libfuse#1053] [General]
In `passthrough_hp.cc`, `do_read` calls function `fuse_reply_data`, but I can not find a way to get the number of bytes it really read.

I have read `fuse_reply_data` and its callees, it's too hard for me to get the number of bytes it really read.

I want to get num of bytes is import for our fuse(modified based on `passthrough_hp.cc`), how can I get it?

**Comment by trapexit** [comment]
Not sure what you mean. If you call fuse_reply_data you are writing to the kernel and it takes the whole amount sent upto the max size.

**Comment by bsbernd** [comment]
@rockeet I guess the code could have been written to return the number of actual bytes having been processed as positive value. Issue is that it would have included fuse-headers - which is probably neither what you want.
So current getting the actual number of read bytes is not possible. Please let me know if you have suggestions how to improve that.

**Comment by rockeet** [comment]
@bsbernd  I've read the code, it's too complex, I'm not sure a positive return value of `fuse_reply_data` is the actual number of read bytes plus fuse-header len, if it is, it is nearly what I want, I just need to subtract the fuse-header len.

A better solution should be add a function like `fuse_reply_data_plus` which returns the actual number of read bytes:
```c
int fuse_reply_data_plus(fuse_req_t req, struct fuse_bufvec *bufv,
        enum fuse_buf_copy_flags flags, ssize_t* num_bytes_actual_read);
```

**Comment by rockeet** [comment]
> Not sure what you mean. If you call fuse_reply_data you are writing to the kernel and it takes the whole amount sent upto the max size.

I'm writing a fuse for messaging, the actual number of read bytes is needed for our fuse.

**Comment by trapexit** [comment]
What do you mean exactly by "actually number of read bytes"? /dev/fuse is not going to take partial messages. The design of all the code is such that it expects that atomic writing behavior. See fuse_send_msg(). There is a writev() and that's it. The return value is not considered for anything but error.

**Comment by bsbernd** [comment]
@trapexit I guess it is something like accounting. And if it is using splice you don't know how much it actually transfers.

@rockeet Feel free to work on it. I'm moving it to discussions, as there is no issue here - the code works as designed.

**Comment by trapexit** [comment]
@bsbernd I'm not saying at all that returning the bytes sent would have been good to have but given the contract (which isn't exactly the most clearly explained) it wasn't strictly necessary. I only just skimmed the code again a few hours ago when I replied but it either errors or ignores a short write. Whether one of those is a problem depends on how the kernel behaves but IIRC it errors if too much is sent or too fails to take it all.

**Comment by rockeet** [comment]
@bsbernd We need this for accounting.
@trapexit I think the actual read size is included in libfuse reply to /dev/fuse, the issue is just it is not returned to libfuse's user.

**Reply by trapexit** [reply]
Why do you need it for accounting? Accounting of what? How many bytes are sent to the kernel? Just add the data size to packet header. If it succeeded you add. If it failed you don't.

**Reply by rockeet** [reply]
In passthrough_hp.cc, `do_read` just passthrough the read request to target file, thus the read result on target file is the expected result to kernel /dev/fuse then to user app side, we do the account when we got the read result on target file. In high level api, the read result is obvious, in low level api as passthrough_hp.cc, it is not obvious. We just want it in low level api.

**Reply by trapexit** [reply]
What do you mean "not obvious"? What read result? The high level API is built on the low level API. If it is possible in the high level it would be possible in the low level. If fuse_reply_data succeeds all your data was sent. If it doesn't succeed it didn't.

**Reply by rockeet** [reply]
In high level api, [xmp_read](https://github.com/libfuse/libfuse/blob/master/example/passthrough.c#L332) calls `pread` which returns the actual read bytes, then returned from user callback(fuse_operations.read) to libfuse, then libfuse reply to kernel /dev/fuse.

low level api reduced the memcpy in read/write thus it is not such obvious, the equivalent operation in `fuse_reply_data` should be `pread`, we just need the equivalent return value of `pread`.

**Reply by trapexit** [reply]
No, we don't need that. Not all files work the same way. FUSE /dev/fuse takes atomic writes. You will not get a short write. It is a fundamental guarantee of the design. If it didn't work that way the libfuse code would be fundamentally broken. It *could* return the bytes sent to /dev/fuse but it is not needed. It succeeds and all bytes are written or it fails and you get -errno. There are no short writes that wouldn't be an error.

**Comment by bsbernd** [comment]
@trapexit I don't think this is about short write, but some accounting (like creating stats how many data have been written/read).

@rockeet If you want to have the result from fuse_reply_data - please write a patch and create a new function. Though we cannot guarantee this will work forever. Eventually we will get zero copy as being currently worked on in ublk - all data handling is then in kernel - no result visible to userspace. I.e. that new function wouldn't work anymore then.

**Reply by trapexit** [reply]
But how many bytes written is implied by the success or failure of the call. You give it a buffer of X bytes it is X + FUSE overhead on success and 0 on failure.

**Reply by bsbernd** [reply]
Your file might be smaller than what struct fuse_lowlevel_ops::read is trying to read. If fuse is the only user - no issue, but there are races - you don't know.

**Reply by trapexit** [reply]
If we are talking about splicing, yes but we have not been told that is the concern. It doesn't hurt to add it for that situation but I wanted to make it clear what the problem was and the setup. That code's behavior is entirely dependent on compile time and runtime facts that we've no details about.

**Comment by bsbernd** [comment]
What I mean is: fuse_lowlevel_ops::read(..., size = X, offset = o, ...), but the file to be read has size Y (with `Y < offset + X`). I.e. fuse_reply_data can only send `Y - offset`.
